src/modeling_library/custom_det_gf.jl

```diff
@@ -0,0 +1,115 @@
+#############################################################
+# abstractions for constructing custom generative functions #
+#############################################################
+
+"""
+    CustomDetGFTrace{T,S} <: Trace
+
+Trace type for custom generative deterministic generative function.
+"""
+struct CustomDetGFTrace{T,S} <: Trace
+    retval::T
+    state::S
+    args::Tuple
+    gen_fn::Any
+end
+
+get_args(trace::CustomDetGFTrace) = trace.args
+
+get_retval(trace::CustomDetGFTrace) = trace.retval
+
+get_choices(trace::CustomDetGFTrace) = EmptyChoiceMap()
+
+get_score(trace::CustomDetGFTrace) = 0.
+
+project(trace::CustomDetGFTrace, selection::Selection) = 0.
+
+get_gen_fn(trace::CustomDetGFTrace) = trace.gen_fn
+
+"""
+    CustomDetGF{T,S} <: GenerativeFunction{T,CustomDetGFTrace{T,S}} 
+
+Abstract type for a custom deterministic generative function. 
+"""
+abstract type CustomDetGF{T,S} <: GenerativeFunction{T,CustomDetGFTrace{T,S}} end
+
+"""
+    retval, state = execute_det(gen_fn::CustomDetGF, args)
+
+Execute the generative function and return the return value and the state.
+"""
+function execute_det end
+
+"""
+    state, retval, retdiff = update_det(gen_fn::CustomDetGF, state, args, argdiffs)
+
+Update the arguments to the generative function and return new return value and state.
+"""
+function update_det end
+
+"""
+    arg_grads = gradient_det(gen_fn::CustomDetGF, state, retgrad) 
+
+Return the gradient tuple with respect to the arguments.
+"""
+function gradient_det end
+
+"""
+    arg_grads = accumulate_param_gradients_get!(trace::CustomDefGF, retgrad, scaler)
+    
+Increment gradient accumulators for parameters the gradient with respect to the
+arguments, optionally scaled, and return the gradient with respect to the
+arguments (not scaled).
+
+Given the previous state and a gradient with respect to the return value \$∇_y
+J\$ (`retgrad`), return the following gradient (`arg_grads`) with respect to
+the arguments \$x\$:
+```math
+∇_x J
+```
+Also increment the gradient accumulators for the trainable parameters \$Θ\$ of
+the function by:
+```math
+s * ∇_Θ J
+```
+where \$s\$ is `scaler`.
```

Comment:
`scaler` is a bit of a tongue twister for me since I so expect `scalar`.  Perhaps we could rename to `scale_factor`?

Reply:
Okay, good idea. This name is used elsewhere in the code, so I'll make a separate issue for that.